httpbp/response.go

```diff
@@ -0,0 +1,218 @@
+	// StatusCode() of the HTTPError.
+	Message string
+
+	// Cause is an optional error that can be used to retain the error that
+	// led to us returning an HTTP error to the client.
+	Cause error
+}
+
+// Error returns the standard error string, this is not returned to the client.
+func (e HTTPError) Error() string {
+	return fmt.Sprintf(
+		"httpbp: http error with code %d, message %q and cause %v",
+		e.Code,
+		e.Message,
+		e.Cause,
+	)
+}
+
+// ResponseMessage returns the error message to send to the client.
+func (e HTTPError) ResponseMessage() string {
+	if e.Message != "" {
+		return e.Message
+	}
+	return http.StatusText(e.StatusCode())
+}
+
+// As implements helper interface for errors.As.
+//
+// If v is pointer to either HTTPError or *HTTPError,
+// *v will be set into this error.
+// Otherwise, As will try errors.As against the Cause of the HTTPError
+func (e HTTPError) As(v interface{}) bool {
+	if target, ok := v.(*HTTPError); ok {
+		*target = e
+		return true
+	} else if target, ok := v.(**HTTPError); ok {
+		*target = &e
+		return true
+	}
+	return errors.As(e.Cause, v)
+}
+
+// Unwrap implements helper interface for errors.Unwrap.  Returns the optional
+// e.Cause error.
+func (e HTTPError) Unwrap() error {
+	return e.Cause
+}
+
+// StatusCode returns the HTTP status code to set on the response.  Defaults to
+// 500 if Code is not set on the HTTPError.
+func (e HTTPError) StatusCode() int {
+	if e.Code == 0 {
+		return http.StatusInternalServerError
+	}
+	return e.Code
+}
+
+var (
+	_ error = HTTPError{}
+	_ error = (*HTTPError)(nil)
+)
+
+// ResponseCookies is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add cookies to the
+// response.
+type ResponseCookies interface {
+	// Return a list of all cookies to set on the response.
+	Cookies() []*http.Cookie
+}
+
+// ResponseHeaders is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically add http headers to the
+// response.
+type ResponseHeaders interface {
+	// Headers returns the http.Header container that can be used to add headers
+	// to the response.
+	Headers() http.Header
+}
+
+// ResponseCode is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically set the response code
+// to something other than 200.
+//
+// If you implement the ErrorResponse interface, any error response codes will
+// be based on the ErrorMapper rather than the StatusCode method implemented in
+// this interface.
+type ResponseCode interface {
+	// StatusCode returns the HTTP status code to set on the response.
+	StatusCode() int
+}
+
+// ErrorResponse is an interface that your Response objects can implement in
+// order to have the httpbp.Encode methods automatically return http errors.
+type ErrorResponse interface {
+	// Err returns the HTTPError set on the response.
+	Err() error
+}
```

Comment:
We likely need to provide an example, either here or at `HTTPError`, to demonstrate how to use `HTTPError` to implement `ErrorResponse` (this does not necessarily need to be a full testable example, it can just be code snippet inside the doc comment).